openmc/surface.py

```diff
@@ -268,62 +332,108 @@ def from_hdf5(group):
             Instance of surface subclass
 
         """
+
         surface_id = int(group.name.split('/')[-1].lstrip('surface '))
         name = group['name'][()].decode() if 'name' in group else ''
         surf_type = group['type'][()].decode()
         bc = group['boundary_type'][()].decode()
         coeffs = group['coefficients'][...]
+        kwargs = {'boundary_type': bc, 'name': name, 'surface_id': surface_id}
 
-        # Create the Surface based on its type
-        if surf_type == 'x-plane':
-            x0 = coeffs[0]
-            surface = XPlane(x0, bc, name, surface_id)
-
-        elif surf_type == 'y-plane':
-            y0 = coeffs[0]
-            surface = YPlane(y0, bc, name, surface_id)
-
-        elif surf_type == 'z-plane':
-            z0 = coeffs[0]
-            surface = ZPlane(z0, bc, name, surface_id)
-
-        elif surf_type == 'plane':
-            A, B, C, D = coeffs
-            surface = Plane(A, B, C, D, bc, name, surface_id)
-
-        elif surf_type == 'x-cylinder':
-            y0, z0, r = coeffs
-            surface = XCylinder(y0, z0, r, bc, name, surface_id)
-
-        elif surf_type == 'y-cylinder':
-            x0, z0, r = coeffs
-            surface = YCylinder(x0, z0, r, bc, name, surface_id)
-
-        elif surf_type == 'z-cylinder':
-            x0, y0, r = coeffs
-            surface = ZCylinder(x0, y0, r, bc, name, surface_id)
-
-        elif surf_type == 'sphere':
-            x0, y0, z0, r = coeffs
-            surface = Sphere(x0, y0, z0, r, bc, name, surface_id)
-
-        elif surf_type in ['x-cone', 'y-cone', 'z-cone']:
-            x0, y0, z0, r2 = coeffs
-            if surf_type == 'x-cone':
-                surface = XCone(x0, y0, z0, r2, bc, name, surface_id)
-            elif surf_type == 'y-cone':
-                surface = YCone(x0, y0, z0, r2, bc, name, surface_id)
-            elif surf_type == 'z-cone':
-                surface = ZCone(x0, y0, z0, r2, bc, name, surface_id)
-
-        elif surf_type == 'quadric':
-            a, b, c, d, e, f, g, h, j, k = coeffs
-            surface = Quadric(a, b, c, d, e, f, g, h, j, k, bc, name, surface_id)
+        cls = _SURFACE_CLASSES[surf_type]
+
+        return cls(*coeffs, **kwargs)
+
+
+class PlaneMixin(metaclass=ABCMeta):
+    """A Plane mixin class for all operations on order 1 surfaces"""
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._periodic_surface = None
+
+    @property
+    def periodic_surface(self):
+        return self._periodic_surface
+
+    @periodic_surface.setter
+    def periodic_surface(self, periodic_surface):
+        check_type('periodic surface', periodic_surface, Plane)
+        self._periodic_surface = periodic_surface
+        periodic_surface._periodic_surface = self
+
+    def _get_base_coeffs(self):
```

Comment:
Should this one just be removed altogether?

Reply:
It's possibly starting to look that way and I'm all for simplifying wherever we can. Since almost every class has `x0, y0, z0` or something close to that translations should be able to be applied very easily without need for abstraction. I think we could even do the rotations for almost all the classes without abstraction since, but it might come back to bite us if we get rid of it. I need to think on this some more...

Reply:
My point with this one is that `_get_base_coeffs` now appears as an `@abstractmethod` on `Surface`, from which all concrete surface classes derive. This method here in `PlaneMixin` is empty and doesn't functionally do anything. The same thing goes for `QuadricMixin`.

Reply:
Ah yes, I see what you're saying, I totally misinterpreted your comment.